chain/network/src/routing/graph_v2/mod.rs

```diff
@@ -0,0 +1,668 @@
+                // V2 routing protocol doesn't broadcast tombstones; don't bother to sign them
+                *edge = Edge::make_fake_edge(peer0, peer1, edge.nonce() + 1);
+            }
+            assert!(edge.edge_type() == EdgeState::Removed);
+        }
+
+        self.edge_cache.remove_tree(peer_id);
+        self.peer_distances.remove(peer_id);
+    }
+
+    /// Handles connection of a new peer or nonce refresh for an existing one.
+    /// - Updates the state of `local_edges`.
+    /// - Adds or updates the nonce in the `edge_cache`.
+    /// - If we don't already have a DistanceVector for this peer, initializes one.
+    pub(crate) fn add_or_update_direct_peer(&mut self, peer_id: PeerId, edge: Edge) -> bool {
+        assert_eq!(edge.edge_type(), EdgeState::Active);
+
+        // We have this nonce or a newer one already; ignore the update entirely
+        if self.edge_cache.has_edge_nonce_or_newer(&edge) {
+            return true;
+        }
+
+        // Reject invalid edge
+        if !self.verify_and_cache_edge_nonces(&vec![edge.clone()]) {
+            return false;
+        }
+
+        // Update the state of `local_edges`
+        self.local_edges.insert(peer_id.clone(), edge.clone());
+
+        // If we don't already have a DistanceVector received from this peer,
+        // create one for it and process it as if we received it
+        if !self.peer_distances.contains_key(&peer_id) {
+            self.handle_distance_vector(&network_protocol::DistanceVector {
+                root: peer_id.clone(),
+                distances: vec![
+                    // The peer has distance 0 to itself
+                    AdvertisedPeerDistance { destination: peer_id, distance: 0 },
+                    // The peer is distance 1 from this node
+                    AdvertisedPeerDistance {
+                        destination: self.config.node_id.clone(),
+                        distance: 1,
+                    },
+                ],
+                edges: vec![edge],
+            });
+        }
+
+        true
+    }
+
+    pub(crate) fn handle_network_change(
+        &mut self,
+        _clock: &time::Clock,
+        update: &NetworkTopologyChange,
+    ) -> bool {
+        match update {
+            NetworkTopologyChange::PeerConnected(peer_id, edge) => {
+                self.add_or_update_direct_peer(peer_id.clone(), edge.clone())
+            }
+            NetworkTopologyChange::PeerDisconnected(peer_id) => {
+                self.remove_direct_peer(peer_id);
+                true
+            }
+            NetworkTopologyChange::PeerAdvertisedDistances(distance_vector) => {
+                self.handle_distance_vector(distance_vector)
+            }
+        }
+    }
+
+    /// Computes and returns "next hops" for all reachable destinations in the network.
+    /// Accepts a set of "unreliable peers" to avoid routing through.
+    /// TODO: Actually avoid the unreliable peers
```

Comment:
Is this planned for this PR?

Reply:
Not for this PR; it is not clear from available documentation why we define unreliable peers as we do ([based on height of their chain](https://github.com/near/nearcore/blob/9c5e42737eb0bcc8a2c2ce667c84b33e3ad3bae0/chain/network/src/peer_manager/peer_manager_actor.rs#L430)) and why we should not route through them. It warrants further investigation, and possibly we will get rid of this concept.